### ISSUE TITLE
Test if the block downloader and verifier error checking is correct

### ISSUE THREAD

**Issue by conradoplg** [ZcashFoundation/zebra#2909]
## Motivation

In https://github.com/ZcashFoundation/zebra/pull/2890/ a downcast was added in the block downloader and verifier, after verifying the block, to differentiate between a verification error and a generic service error (since the buffered verification service returns a `BoxError`). Depending on the downcast result the underlying error is put into one of the `BlockDownloadVerifyError` variants.

The error returned is eventually used in `should_restart_sync` by the syncer to ignore specific errors when deciding whether to restart the sync or not.

However, if the downcast is wrong (or becomes wrong after some refactoring, for example) then this filter will fail because all errors will be considered "service errors" which will make the syncer restart everytime one of these "ignorable" errors happen, which slows the sync down.

https://github.com/ZcashFoundation/zebra/pull/2890/ added a best-effort check which logs an error if it detects that happening. However, it would be best to have a proper test for it, however it is tricky to write and was postponed.

In other words, what needs to be tested is: when these errors happen, are they correctly detected by `should_restart_sync`?

**Comment by mpguerra** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Engineering-Team-5fb24d9264a3e8000e666a9e/issues/ZcashFoundation/zebra/2909?planning-poker) with ZenHub @conradoplg @dconnolly @jvff @oxarbitrage @teor2345 @upbqdn

**Comment by teor2345** [comment]
We should fix the actual bug risk in #2908 instead.
